Material Canvas: Remove material hot reload notification bus from material system classes and assets

gadams3 opened the pull request:
## What does this PR do?

Material editor, material canvas, material instance editor, and thumbnails all use the material component to load, reload, and display material previews in their viewports. These systems exercise material hot reloading more than any others besides ASV samples and tests.

Changes were previously made to reenable material finalization by default in the material builder. That allowed systems that manage materials to be able to rely on standard asset ready and reload notifications. The material component uses the asset bus for more conventional loading and hot reloading at the high level. It requires management and access to the most recent version of material assets to enumerate, get, and set materials and properties.

Other changes were also made for the instance database to be able to create and differentiate instances for different versions of assets with the same asset ID.

Those changes unblocked several issues but some still remain.

With the material reload notification bus logic still in place, competing systems are attempting to load materials and shaders.

Anytime material inputs are changed in a material graph, the material type and its assets need to be reprocessed, which sometimes results in asserts and errors related to shader resource group and data comparisons.

I suspect that those issues are caused by the instance database. The material, material type, and shaders may reload correctly and send the correct asset reload notifications but the instance database could be holding on to old versions of shaders and block new versions from being created when requested.

Removing the material reload notification bus and the self-reloading of materials, material types, and material assets has stopped most of the problems, while still preserving material hot reloading in the material component.

Another change was recently made to the material assignment class. This change sends a notification every time material instances are rebuilt. This code was crashing because it dereferenced a null instance pointer. The material instance pointer was null because the instance database function failed to create an instance, likely due to a collision with a pre-existing random ID.

The terrain feature processor would need to switch to conventional asset notification handling to hot reload materials.

Corresponding sample viewer changes https://github.com/o3de/o3de-atom-sampleviewer/pull/550

Signed-off-by: gadams3 <guthadam@amazon.com>

Tested by iterating on materials using material canvas and seeing that there were no asserts or log spam.
Tested by opening terrain test levels in automated testing project, starting simulation and game mode, launching material editor with autosave enabled, editing the materials assigned to terrain, and seeing the changes reload in the main editor.

gadams3 commented:
Occasional logs pre-changes...

```
==================================================================
System: Trace::Assert
 D:/projects/o3de/Gems/Atom/RHI/Code/Source/RHI.Reflect/ConstantsLayout.cpp(126): (15892) 'bool __cdecl AZ::RHI::ConstantsLayout::ValidateAccess(struct AZ::RHI::Handle<unsigned int,class AZ::RHI::ShaderInputConstantDescriptor>) const'
System: Inline Constant Input index '1' out of range [0,1).
System: ------------------------------------------------
System: D:\projects\o3de\Gems\Atom\RHI\Code\Source\RHI.Reflect\ConstantsLayout.cpp (127) : AZ::RHI::ConstantsLayout::ValidateAccess
System: D:\projects\o3de\Gems\Atom\RHI\Code\Source\RHI\ConstantsData.cpp (31) : AZ::RHI::ConstantsData::ValidateConstantAccess
System: D:\projects\o3de\Gems\Atom\RHI\Code\Source\RHI\ConstantsData.cpp (281) : AZ::RHI::ConstantsData::SetConstant<AZ::Color>
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (571) : AZ::RPI::Material::SetPropertyValue<AZ::Vector4>
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (125) : AZ::RPI::Material::Init
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (255) : AZ::RPI::Material::OnAssetReloaded
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (272) : AZ::RPI::Material::OnMaterialAssetReinitialized
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (146) : AZ::Internal::EBusContainer<AZ::RPI::MaterialReloadNotifications,AZ::RPI::MaterialReloadNotifications,1,1>::Dispatcher<AZ::EBus<AZ::RPI::MaterialReloadNotifications,AZ::RP
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Reflect\Material\MaterialAsset.cpp (263) : AZ::RPI::MaterialAsset::OnMaterialTypeAssetReinitialized
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (146) : AZ::Internal::EBusContainer<AZ::RPI::MaterialReloadNotifications,AZ::RPI::MaterialReloadNotifications,1,1>::Dispatcher<AZ::EBus<AZ::RPI::MaterialReloadNotifications,AZ::RP
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Reflect\Material\MaterialTypeAsset.cpp (218) : AZ::RPI::MaterialTypeAsset::ReinitializeAsset
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Reflect\Material\MaterialTypeAsset.cpp (225) : AZ::RPI::MaterialTypeAsset::OnAssetReloaded
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (146) : AZ::Internal::EBusContainer<AZ::Data::AssetEvents,AZ::Data::AssetEvents,1,1>::Dispatcher<AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents> >::Event<void (__cdecl AZ::D
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (1689) : AZ::Data::AssetManager::AssignAssetData
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (1852) : AZ::Data::AssetManager::NotifyAssetReloaded
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\std\function\function_template.h (148) : AZStd::Internal::function_util::get_invoker<void __cdecl(void),AZStd::Internal::function_util::function_obj_tag,AZ::Internal::EBusEnvironmentAllocator>::call<`AZ::BusI
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Policies.h (267) : AZ::EBusQueuePolicy<1,AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents>,AZStd::recursive_mutex>::Execute
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (591) : AZ::Data::AssetManager::DispatchEvents
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (1360) : AZ::Internal::EBusContainer<AZ::SystemTickEvents,AZ::SystemTickEvents,0,1>::Dispatcher<AZ::EBus<AZ::SystemTickEvents,AZ::SystemTickEvents> >::Broadcast<void (__cdecl AZ::
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Component\ComponentApplication.cpp (1519) : AZ::ComponentApplication::TickSystem
System: D:\projects\o3de\Gems\Atom\Tools\AtomToolsFramework\Code\Source\Application\AtomToolsApplication.cpp (286) : AtomToolsFramework::AtomToolsApplication::OnIdle
System: 00007FF8D0BEF72A (Qt5Core) : QTimer::timeout
System: 00007FF8D0BE82DD (Qt5Core) : QObject::event
System: 00007FF8D2BE49A1 (Qt5Widgets) : QApplicationPrivate::notify_helper
System: 00007FF8D2BE3A23 (Qt5Widgets) : QApplication::notify
System: ==================================================================
System:
==================================================================
System: Trace::Assert
 D:/projects/o3de/Gems/Atom/RPI/Code/Source/RPI.Reflect/Material/MaterialAsset.cpp(213): (15892) 'const class AZStd::vector<class AZ::RPI::MaterialPropertyValue,class AZStd::allocator> &__cdecl AZ::RPI::MaterialAsset::GetPropertyValues(void)'
System: MaterialAsset should be finalized but does not have the right number of property values.
System: ------------------------------------------------
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Reflect\Material\MaterialAsset.cpp (216) : AZ::RPI::MaterialAsset::GetPropertyValues
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (118) : AZ::RPI::Material::Init
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (255) : AZ::RPI::Material::OnAssetReloaded
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (272) : AZ::RPI::Material::OnMaterialAssetReinitialized
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (146) : AZ::Internal::EBusContainer<AZ::RPI::MaterialReloadNotifications,AZ::RPI::MaterialReloadNotifications,1,1>::Dispatcher<AZ::EBus<AZ::RPI::MaterialReloadNotifications,AZ::RP
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Reflect\Material\MaterialAsset.cpp (326) : AZ::RPI::MaterialAsset::ReinitializeMaterialTypeAsset
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Reflect\Material\MaterialAsset.cpp (334) : AZ::RPI::MaterialAsset::OnAssetReloaded
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (146) : AZ::Internal::EBusContainer<AZ::Data::AssetEvents,AZ::Data::AssetEvents,1,1>::Dispatcher<AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents> >::Event<void (__cdecl AZ::D
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (1689) : AZ::Data::AssetManager::AssignAssetData
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (1852) : AZ::Data::AssetManager::NotifyAssetReloaded
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\std\function\function_template.h (148) : AZStd::Internal::function_util::get_invoker<void __cdecl(void),AZStd::Internal::function_util::function_obj_tag,AZ::Internal::EBusEnvironmentAllocator>::call<`AZ::BusI
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Policies.h (267) : AZ::EBusQueuePolicy<1,AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents>,AZStd::recursive_mutex>::Execute
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (591) : AZ::Data::AssetManager::DispatchEvents
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (1360) : AZ::Internal::EBusContainer<AZ::SystemTickEvents,AZ::SystemTickEvents,0,1>::Dispatcher<AZ::EBus<AZ::SystemTickEvents,AZ::SystemTickEvents> >::Broadcast<void (__cdecl AZ::
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Component\ComponentApplication.cpp (1519) : AZ::ComponentApplication::TickSystem
System: D:\projects\o3de\Gems\Atom\Tools\AtomToolsFramework\Code\Source\Application\AtomToolsApplication.cpp (286) : AtomToolsFramework::AtomToolsApplication::OnIdle
System: 00007FF8D0BEF72A (Qt5Core) : QTimer::timeout
System: 00007FF8D0BE82DD (Qt5Core) : QObject::event
System: 00007FF8D2BE49A1 (Qt5Widgets) : QApplicationPrivate::notify_helper
System: 00007FF8D2BE3A23 (Qt5Widgets) : QApplication::notify
System: 00007FF8D0BC17CB (Qt5Core) : QCoreApplication::notifyInternal2
System: 00007FF8D0C0C1B9 (Qt5Core) : QEventDispatcherWin32Private::sendTimerEvent
System: 00007FF8D0C0A5E9 (Qt5Core) : QEventDispatcherWin32::event
System: 00007FF8D2BE49A1 (Qt5Widgets) : QApplicationPrivate::notify_helper
System: 00007FF8D2BE3A23 (Qt5Widgets) : QApplication::notify
System: ==================================================================
System:
==================================================================
System: Trace::Assert
 D:/projects/o3de/Gems/Atom/RPI/Code/Source/RPI.Public/Material/Material.cpp(119): (15892) 'enum AZ::RHI::ResultCode __cdecl AZ::RPI::Material::Init(class AZ::RPI::MaterialAsset &)'
System: The number of properties in this material doesn't match the property layout
System: ------------------------------------------------
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (121) : AZ::RPI::Material::Init
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (255) : AZ::RPI::Material::OnAssetReloaded
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (272) : AZ::RPI::Material::OnMaterialAssetReinitialized
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (146) : AZ::Internal::EBusContainer<AZ::RPI::MaterialReloadNotifications,AZ::RPI::MaterialReloadNotifications,1,1>::Dispatcher<AZ::EBus<AZ::RPI::MaterialReloadNotifications,AZ::RP
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Reflect\Material\MaterialAsset.cpp (326) : AZ::RPI::MaterialAsset::ReinitializeMaterialTypeAsset
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Reflect\Material\MaterialAsset.cpp (334) : AZ::RPI::MaterialAsset::OnAssetReloaded
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (146) : AZ::Internal::EBusContainer<AZ::Data::AssetEvents,AZ::Data::AssetEvents,1,1>::Dispatcher<AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents> >::Event<void (__cdecl AZ::D
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (1689) : AZ::Data::AssetManager::AssignAssetData
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (1852) : AZ::Data::AssetManager::NotifyAssetReloaded
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\std\function\function_template.h (148) : AZStd::Internal::function_util::get_invoker<void __cdecl(void),AZStd::Internal::function_util::function_obj_tag,AZ::Internal::EBusEnvironmentAllocator>::call<`AZ::BusI
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Policies.h (267) : AZ::EBusQueuePolicy<1,AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents>,AZStd::recursive_mutex>::Execute
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (591) : AZ::Data::AssetManager::DispatchEvents
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (1360) : AZ::Internal::EBusContainer<AZ::SystemTickEvents,AZ::SystemTickEvents,0,1>::Dispatcher<AZ::EBus<AZ::SystemTickEvents,AZ::SystemTickEvents> >::Broadcast<void (__cdecl AZ::
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Component\ComponentApplication.cpp (1519) : AZ::ComponentApplication::TickSystem
System: D:\projects\o3de\Gems\Atom\Tools\AtomToolsFramework\Code\Source\Application\AtomToolsApplication.cpp (286) : AtomToolsFramework::AtomToolsApplication::OnIdle
System: 00007FF8D0BEF72A (Qt5Core) : QTimer::timeout
System: 00007FF8D0BE82DD (Qt5Core) : QObject::event
System: 00007FF8D2BE49A1 (Qt5Widgets) : QApplicationPrivate::notify_helper
System: 00007FF8D2BE3A23 (Qt5Widgets) : QApplication::notify
System: 00007FF8D0BC17CB (Qt5Core) : QCoreApplication::notifyInternal2
System: 00007FF8D0C0C1B9 (Qt5Core) : QEventDispatcherWin32Private::sendTimerEvent
System: 00007FF8D0C0A5E9 (Qt5Core) : QEventDispatcherWin32::event
System: 00007FF8D2BE49A1 (Qt5Widgets) : QApplicationPrivate::notify_helper
System: 00007FF8D2BE3A23 (Qt5Widgets) : QApplication::notify
System: 00007FF8D0BC17CB (Qt5Core) : QCoreApplication::notifyInternal2
System: ==================================================================
System:
==================================================================
System: Trace::Assert
 D:/projects/o3de/Gems/Atom/RPI/Code/Source/RPI.Reflect/Material/MaterialAsset.cpp(213): (15892) 'const class AZStd::vector<class AZ::RPI::MaterialPropertyValue,class AZStd::allocator> &__cdecl AZ::RPI::MaterialAsset::GetPropertyValues(void)'
System: MaterialAsset should be finalized but does not have the right number of property values.
System: ------------------------------------------------
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Reflect\Material\MaterialAsset.cpp (216) : AZ::RPI::MaterialAsset::GetPropertyValues
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (121) : AZ::RPI::Material::Init
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (255) : AZ::RPI::Material::OnAssetReloaded
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (272) : AZ::RPI::Material::OnMaterialAssetReinitialized
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (146) : AZ::Internal::EBusContainer<AZ::RPI::MaterialReloadNotifications,AZ::RPI::MaterialReloadNotifications,1,1>::Dispatcher<AZ::EBus<AZ::RPI::MaterialReloadNotifications,AZ::RP
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Reflect\Material\MaterialAsset.cpp (326) : AZ::RPI::MaterialAsset::ReinitializeMaterialTypeAsset
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Reflect\Material\MaterialAsset.cpp (334) : AZ::RPI::MaterialAsset::OnAssetReloaded
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (146) : AZ::Internal::EBusContainer<AZ::Data::AssetEvents,AZ::Data::AssetEvents,1,1>::Dispatcher<AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents> >::Event<void (__cdecl AZ::D
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (1689) : AZ::Data::AssetManager::AssignAssetData
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (1852) : AZ::Data::AssetManager::NotifyAssetReloaded
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\std\function\function_template.h (148) : AZStd::Internal::function_util::get_invoker<void __cdecl(void),AZStd::Internal::function_util::function_obj_tag,AZ::Internal::EBusEnvironmentAllocator>::call<`AZ::BusI
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Policies.h (267) : AZ::EBusQueuePolicy<1,AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents>,AZStd::recursive_mutex>::Execute
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (591) : AZ::Data::AssetManager::DispatchEvents
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (1360) : AZ::Internal::EBusContainer<AZ::SystemTickEvents,AZ::SystemTickEvents,0,1>::Dispatcher<AZ::EBus<AZ::SystemTickEvents,AZ::SystemTickEvents> >::Broadcast<void (__cdecl AZ::
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Component\ComponentApplication.cpp (1519) : AZ::ComponentApplication::TickSystem
System: D:\projects\o3de\Gems\Atom\Tools\AtomToolsFramework\Code\Source\Application\AtomToolsApplication.cpp (286) : AtomToolsFramework::AtomToolsApplication::OnIdle
System: 00007FF8D0BEF72A (Qt5Core) : QTimer::timeout
System: 00007FF8D0BE82DD (Qt5Core) : QObject::event
System: 00007FF8D2BE49A1 (Qt5Widgets) : QApplicationPrivate::notify_helper
System: 00007FF8D2BE3A23 (Qt5Widgets) : QApplication::notify
System: 00007FF8D0BC17CB (Qt5Core) : QCoreApplication::notifyInternal2
System: 00007FF8D0C0C1B9 (Qt5Core) : QEventDispatcherWin32Private::sendTimerEvent
System: 00007FF8D0C0A5E9 (Qt5Core) : QEventDispatcherWin32::event
System: 00007FF8D2BE49A1 (Qt5Widgets) : QApplicationPrivate::notify_helper
System: 00007FF8D2BE3A23 (Qt5Widgets) : QApplication::notify
System: ==================================================================
System:
==================================================================
System: Trace::Assert
 D:/projects/o3de/Gems/Atom/RPI/Code/Source/RPI.Reflect/Material/MaterialAsset.cpp(213): (15892) 'const class AZStd::vector<class AZ::RPI::MaterialPropertyValue,class AZStd::allocator> &__cdecl AZ::RPI::MaterialAsset::GetPropertyValues(void)'
System: MaterialAsset should be finalized but does not have the right number of property values.
System: ------------------------------------------------
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Reflect\Material\MaterialAsset.cpp (216) : AZ::RPI::MaterialAsset::GetPropertyValues
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (123) : AZ::RPI::Material::Init
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (255) : AZ::RPI::Material::OnAssetReloaded
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (272) : AZ::RPI::Material::OnMaterialAssetReinitialized
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (146) : AZ::Internal::EBusContainer<AZ::RPI::MaterialReloadNotifications,AZ::RPI::MaterialReloadNotifications,1,1>::Dispatcher<AZ::EBus<AZ::RPI::MaterialReloadNotifications,AZ::RP
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Reflect\Material\MaterialAsset.cpp (326) : AZ::RPI::MaterialAsset::ReinitializeMaterialTypeAsset
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Reflect\Material\MaterialAsset.cpp (334) : AZ::RPI::MaterialAsset::OnAssetReloaded
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (146) : AZ::Internal::EBusContainer<AZ::Data::AssetEvents,AZ::Data::AssetEvents,1,1>::Dispatcher<AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents> >::Event<void (__cdecl AZ::D
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (1689) : AZ::Data::AssetManager::AssignAssetData
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (1852) : AZ::Data::AssetManager::NotifyAssetReloaded
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\std\function\function_template.h (148) : AZStd::Internal::function_util::get_invoker<void __cdecl(void),AZStd::Internal::function_util::function_obj_tag,AZ::Internal::EBusEnvironmentAllocator>::call<`AZ::BusI
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Policies.h (267) : AZ::EBusQueuePolicy<1,AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents>,AZStd::recursive_mutex>::Execute
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (591) : AZ::Data::AssetManager::DispatchEvents
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (1360) : AZ::Internal::EBusContainer<AZ::SystemTickEvents,AZ::SystemTickEvents,0,1>::Dispatcher<AZ::EBus<AZ::SystemTickEvents,AZ::SystemTickEvents> >::Broadcast<void (__cdecl AZ::
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Component\ComponentApplication.cpp (1519) : AZ::ComponentApplication::TickSystem
System: D:\projects\o3de\Gems\Atom\Tools\AtomToolsFramework\Code\Source\Application\AtomToolsApplication.cpp (286) : AtomToolsFramework::AtomToolsApplication::OnIdle
System: 00007FF8D0BEF72A (Qt5Core) : QTimer::timeout
System: 00007FF8D0BE82DD (Qt5Core) : QObject::event
System: 00007FF8D2BE49A1 (Qt5Widgets) : QApplicationPrivate::notify_helper
System: 00007FF8D2BE3A23 (Qt5Widgets) : QApplication::notify
System: 00007FF8D0BC17CB (Qt5Core) : QCoreApplication::notifyInternal2
System: 00007FF8D0C0C1B9 (Qt5Core) : QEventDispatcherWin32Private::sendTimerEvent
System: 00007FF8D0C0A5E9 (Qt5Core) : QEventDispatcherWin32::event
System: 00007FF8D2BE49A1 (Qt5Widgets) : QApplicationPrivate::notify_helper
System: 00007FF8D2BE3A23 (Qt5Widgets) : QApplication::notify
System: ==================================================================
System:
==================================================================
System: Trace::Assert
 D:/projects/o3de/Gems/Atom/RPI/Code/Source/RPI.Reflect/Material/MaterialAsset.cpp(213): (15892) 'const class AZStd::vector<class AZ::RPI::MaterialPropertyValue,class AZStd::allocator> &__cdecl AZ::RPI::MaterialAsset::GetPropertyValues(void)'
System: MaterialAsset should be finalized but does not have the right number of property values.
System: ------------------------------------------------
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Reflect\Material\MaterialAsset.cpp (216) : AZ::RPI::MaterialAsset::GetPropertyValues
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (121) : AZ::RPI::Material::Init
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (255) : AZ::RPI::Material::OnAssetReloaded
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (272) : AZ::RPI::Material::OnMaterialAssetReinitialized
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (146) : AZ::Internal::EBusContainer<AZ::RPI::MaterialReloadNotifications,AZ::RPI::MaterialReloadNotifications,1,1>::Dispatcher<AZ::EBus<AZ::RPI::MaterialReloadNotifications,AZ::RP
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Reflect\Material\MaterialAsset.cpp (326) : AZ::RPI::MaterialAsset::ReinitializeMaterialTypeAsset
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Reflect\Material\MaterialAsset.cpp (334) : AZ::RPI::MaterialAsset::OnAssetReloaded
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (146) : AZ::Internal::EBusContainer<AZ::Data::AssetEvents,AZ::Data::AssetEvents,1,1>::Dispatcher<AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents> >::Event<void (__cdecl AZ::D
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (1689) : AZ::Data::AssetManager::AssignAssetData
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (1852) : AZ::Data::AssetManager::NotifyAssetReloaded
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\std\function\function_template.h (148) : AZStd::Internal::function_util::get_invoker<void __cdecl(void),AZStd::Internal::function_util::function_obj_tag,AZ::Internal::EBusEnvironmentAllocator>::call<`AZ::BusI
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Policies.h (267) : AZ::EBusQueuePolicy<1,AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents>,AZStd::recursive_mutex>::Execute
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (591) : AZ::Data::AssetManager::DispatchEvents
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (1360) : AZ::Internal::EBusContainer<AZ::SystemTickEvents,AZ::SystemTickEvents,0,1>::Dispatcher<AZ::EBus<AZ::SystemTickEvents,AZ::SystemTickEvents> >::Broadcast<void (__cdecl AZ::
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Component\ComponentApplication.cpp (1519) : AZ::ComponentApplication::TickSystem
System: D:\projects\o3de\Gems\Atom\Tools\AtomToolsFramework\Code\Source\Application\AtomToolsApplication.cpp (286) : AtomToolsFramework::AtomToolsApplication::OnIdle
System: 00007FF8D0BEF72A (Qt5Core) : QTimer::timeout
System: 00007FF8D0BE82DD (Qt5Core) : QObject::event
System: 00007FF8D2BE49A1 (Qt5Widgets) : QApplicationPrivate::notify_helper
System: 00007FF8D2BE3A23 (Qt5Widgets) : QApplication::notify
System: 00007FF8D0BC17CB (Qt5Core) : QCoreApplication::notifyInternal2
System: 00007FF8D0C0C1B9 (Qt5Core) : QEventDispatcherWin32Private::sendTimerEvent
System: 00007FF8D0C0A5E9 (Qt5Core) : QEventDispatcherWin32::event
System: 00007FF8D2BE49A1 (Qt5Widgets) : QApplicationPrivate::notify_helper
System: 00007FF8D2BE3A23 (Qt5Widgets) : QApplication::notify
System: ==================================================================
System:
==================================================================
System: Trace::Assert
 D:/projects/o3de/Gems/Atom/RPI/Code/Source/RPI.Reflect/Material/MaterialAsset.cpp(213): (15892) 'const class AZStd::vector<class AZ::RPI::MaterialPropertyValue,class AZStd::allocator> &__cdecl AZ::RPI::MaterialAsset::GetPropertyValues(void)'
System: MaterialAsset should be finalized but does not have the right number of property values.
System: ------------------------------------------------
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Reflect\Material\MaterialAsset.cpp (216) : AZ::RPI::MaterialAsset::GetPropertyValues
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (123) : AZ::RPI::Material::Init
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (255) : AZ::RPI::Material::OnAssetReloaded
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Material\Material.cpp (272) : AZ::RPI::Material::OnMaterialAssetReinitialized
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (146) : AZ::Internal::EBusContainer<AZ::RPI::MaterialReloadNotifications,AZ::RPI::MaterialReloadNotifications,1,1>::Dispatcher<AZ::EBus<AZ::RPI::MaterialReloadNotifications,AZ::RP
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Reflect\Material\MaterialAsset.cpp (326) : AZ::RPI::MaterialAsset::ReinitializeMaterialTypeAsset
System: D:\projects\o3de\Gems\Atom\RPI\Code\Source\RPI.Reflect\Material\MaterialAsset.cpp (334) : AZ::RPI::MaterialAsset::OnAssetReloaded
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (146) : AZ::Internal::EBusContainer<AZ::Data::AssetEvents,AZ::Data::AssetEvents,1,1>::Dispatcher<AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents> >::Event<void (__cdecl AZ::D
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (1689) : AZ::Data::AssetManager::AssignAssetData
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (1852) : AZ::Data::AssetManager::NotifyAssetReloaded
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\std\function\function_template.h (148) : AZStd::Internal::function_util::get_invoker<void __cdecl(void),AZStd::Internal::function_util::function_obj_tag,AZ::Internal::EBusEnvironmentAllocator>::call<`AZ::BusI
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Policies.h (267) : AZ::EBusQueuePolicy<1,AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents>,AZStd::recursive_mutex>::Execute
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (591) : AZ::Data::AssetManager::DispatchEvents
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (1360) : AZ::Internal::EBusContainer<AZ::SystemTickEvents,AZ::SystemTickEvents,0,1>::Dispatcher<AZ::EBus<AZ::SystemTickEvents,AZ::SystemTickEvents> >::Broadcast<void (__cdecl AZ::
System: D:\projects\o3de\Code\Framework\AzCore\AzCore\Component\ComponentApplication.cpp (1519) : AZ::ComponentApplication::TickSystem
System: D:\projects\o3de\Gems\Atom\Tools\AtomToolsFramework\Code\Source\Application\AtomToolsApplication.cpp (286) : AtomToolsFramework::AtomToolsApplication::OnIdle
System: 00007FF8D0BEF72A (Qt5Core) : QTimer::timeout
System: 00007FF8D0BE82DD (Qt5Core) : QObject::event
System: 00007FF8D2BE49A1 (Qt5Widgets) : QApplicationPrivate::notify_helper
System: 00007FF8D2BE3A23 (Qt5Widgets) : QApplication::notify
System: 00007FF8D0BC17CB (Qt5Core) : QCoreApplication::notifyInternal2
System: 00007FF8D0C0C1B9 (Qt5Core) : QEventDispatcherWin32Private::sendTimerEvent
System: 00007FF8D0C0A5E9 (Qt5Core) : QEventDispatcherWin32::event
System: 00007FF8D2BE49A1 (Qt5Widgets) : QApplicationPrivate::notify_helper
System: 00007FF8D2BE3A23 (Qt5Widgets) : QApplication::notify
System: ==================================================================
Material:
==================================================================
Material: Trace::Warning
 D:/projects/o3de/Gems/Atom/RPI/Code/Source/RPI.Public/Material/Material.cpp(451): 'bool __cdecl AZ::RPI::Material::ValidatePropertyAccess<class AZ::Vector4>(const class AZ::RPI::MaterialPropertyDescriptor *) const'
Material: MaterialPropertyDescriptor is null
Material: ==================================================================
```

invertednormal commented on Code/Framework/AtomCore/AtomCore/Instance/InstanceDatabase.h:
```diff
@@ -362,13 +362,19 @@ namespace AZ
         Data::Instance<Type> InstanceDatabase<Type>::Create(const Asset<AssetData>& asset, const AZStd::any* param)
         {
-            return FindOrCreate(Data::InstanceId::CreateRandom(), asset, param);
+            // Handle the rare case where a random ID might already be in use
+            auto id = Data::InstanceId::CreateRandom();
+            while (Find(id))
+            {
+                id = Data::InstanceId::CreateRandom();
+            }
+            return FindOrCreate(id, asset, param);
```
While it'd be incredibly unlikely to happen, you'd need a mutex here to be completely safe. Otherwise, it may be possible for a different thread to pick the same random id and pass the `Find(id)` check.

VickyAtAZ reviewed (commented):
Does the MaterialTypeAsset change trigger MaterialAsset rebuild? 
I check the the hot-reloading code in MaterialComponentController and it seems only handle MaterialAsset reload. I wonder if that's sufficient to handle all the hot-reload uses cases. 

I am okay with the idea of removing the self hot-reload handling had issues and didn't work well.

invertednormal reviewed (approved):
Looks like a good change overall, appreciate the reduced complexity. Just need to fix up a few places to not break functionality for the final PR.

invertednormal commented on Code/Framework/AtomCore/AtomCore/Instance/InstanceDatabase.h:
```diff
@@ -362,13 +362,19 @@ namespace AZ
         Data::Instance<Type> InstanceDatabase<Type>::Create(const Asset<AssetData>& asset, const AZStd::any* param)
         {
-            return FindOrCreate(Data::InstanceId::CreateRandom(), asset, param);
+            // Handle the rare case where a random ID might already be in use
+            auto id = Data::InstanceId::CreateRandom();
+            while (Find(id))
+            {
+                id = Data::InstanceId::CreateRandom();
+            }
```
Ids are 128 bits, which means assuming perfect randomness, the probably of a collision hits 50% after 2^64 numbers have been generated (18 quintillion). That's a lot, so it's very unlikely a collision will ever happen.

gadams3 commented:
> Does the MaterialTypeAsset change trigger MaterialAsset rebuild? I check the the hot-reloading code in MaterialComponentController and it seems only handle MaterialAsset reload. I wonder if that's sufficient to handle all the hot-reload uses cases.
> 
> I am okay with the idea of removing the self hot-reload handling had issues and didn't work well.

Yes, as long as the asset dependencies are set up correctly, reprocessing shaders and functors should cause material types to be reprocessed, which should also cause materials to be reprocessed. This was the original behavior. 

The problem was that, early on, and with subsequent shader refactors, people were making frequent changes to shaders and materials in common features, like standard PBR, and experiencing exceedingly long build times because of shader compilation and all of the other dependencies. The model builder forces a dependency on standard PBR, which causes all models to be reprocessed when the material type or shaders change. 

To resolve those problems for people making frequent changes to common materials, really just standard PBR, a flag was added to the material builder called finalize materials. With the flag enabled, the builder respects the normal dependency chain described above. With the flag disabled, the material builder breaks the dependency on material types and materials attempt to resolve everything at runtime. If the flag is disabled, you would also have to listen for material type asset notifications, which is part of what the deleted code was doing. The flag is currently enabled by default, meaning changing shaders or material types will also update and send notifications for the material. 

The material component is using the standard asset notifications and all of that seems to work correctly with finalization enabled. However, without this change, all of the other handling and buses are still operating in parallel, causing things to get out of sync, triggering the asserts.

Other work is being done to break the dependency between the model builder and standard PBR. Part of that work should include removing the finalization flag and always respecting the asset dependencies.

gadams3 commented:
The last push updates the terrain materialist component and detail material manager to support hot loading using asset bus notifications.

The terrain material list component logic was not immediately clear because of several conditionals in one function to handle all of the material state changes. However, that function was only called when assets changed or the component deactivated. Following that logic, I did some simplification to update the material instances on the default material and the surface materials when assets load, reload, or the component deactivates. 

The other case is in the original function would probably never occur unless data lingered between editor component activation and deactivation or the containers were changed through script.

Remapping surface tags there's no handle on the ticket bus which will catch editor and script changes. Editor changes should probably be going through normal change notification on the component or configuration edit context but that wasn't being done before.

There was another issue where the terrene material manager was blocking material changes if the change ID did not change since the last time it was added. The change ID keeps count of the number of times properties change and the material is compiled. So the material change ID won't work for that kind of comparison because it will always be the same loading or reloading the same material or another material with the same number of properties. Because the terrain material this component does not provide and API for manipulating material properties, there is no need to check the change ID. That was removed as the last thing preventing hot reloading.

In any event, this correctly supports loading and reloading materials. I do not see the shapes or terrain in the editor unless I am in simulation or game mode. But once in either mode, I can make changes to the materials through the material editor with autosave enabled and see the changes reflected on the terrain.

santorac commented:
> To resolve those problems for people making frequent changes to common materials, really just standard PBR, a flag was added to the material builder called finalize materials. With the flag enabled, the builder respects the normal dependency chain described above. With the flag disabled, the material builder breaks the dependency on material types and materials attempt to resolve everything at runtime. If the flag is disabled, you would also have to listen for material type asset notifications, which is part of what the deleted code was doing. The flag is currently enabled by default, meaning changing shaders or material types will also update and send notifications for the material.

Just one thing to clarify: IIRC The material reload notification bus was not added to support the finalize flag, it actually predates the finalize flag. We needed the notification bus because the asset system did not guarantee the order in which OnAssetReloaded events would be sent. If a shader was changed, it would be reprocessed followed by materialtype, followed by material and fbx. However, the asset manager could call OnAssetReloaded in any order. If it was called for the material first and then for the materialtype, then it would end up using a stale material and you'd still see the old material. 

So it sounds like the asset management system has been improved to guarantee that the order of AssetBus callbacks now follows the order in which the dependent assets are processed.

santorac commented on Gems/Atom/RPI/Code/Source/RPI.Public/Material/Material.cpp:
```diff
@@ -280,7 +242,7 @@ namespace AZ
             // just bumping the m_currentChangeId or some other minor updates. But it's pretty hard to know what exactly needs to be
             // updated to correctly handle the reload, so it's safer to just reinitialize the whole material.
-            OnAssetReloaded(m_materialAsset);
+            Init(*m_materialAsset);
```
Hm, it's interesting that you were able to remove the material reload notification bus, but this is still here. Are you planning to remove this as well?
Though I expect it won't work until this is fixed: https://github.com/o3de/o3de/issues/13188
